File: Lab04/trapecioPool.go
package main

import (
	"fmt"
	"sync"
	"time"
)

type Funcion interface {
	Evaluar(x float64) float64
}

type FuncionCuadratica struct{}

func (f FuncionCuadratica) Evaluar(x float64) float64 {
	return 2.0*x*x + 3.0*x + 0.5
}

type IntegradorTrapecioParalelo struct {
	a, b     float64
	n        int64
	numHilos int
	funcion  Funcion
}

func (it IntegradorTrapecioParalelo) integrarParcial(h float64, inicio, fin int64, wg *sync.WaitGroup, resultado *float64, mutex *sync.Mutex) {
	defer wg.Done()

	suma := 0.0
	for i := inicio; i <= fin; i++ {
		x := it.a + float64(i)*h
		suma += it.funcion.Evaluar(x)
	}

	mutex.Lock()
	*resultado += suma
	mutex.Unlock()
}

func (it IntegradorTrapecioParalelo) Integrar() float64 {
	h := (it.b - it.a) / float64(it.n)
	sumaTotal := 0.5 * (it.funcion.Evaluar(it.a) + it.funcion.Evaluar(it.b))

	var resultado float64 = 0.0
	var wg sync.WaitGroup
	var mutex sync.Mutex

	pasosPorHilo := it.n / int64(it.numHilos)

	for i := 0; i < it.numHilos; i++ {
		inicio := int64(i)*pasosPorHilo + 1
		var fin int64
		if i == it.numHilos-1 {
			fin = it.n - 1
		} else {
			fin = (int64(i+1) * pasosPorHilo)
		}

		wg.Add(1)
		go it.integrarParcial(h, inicio, fin, &wg, &resultado, &mutex)
	}

	wg.Wait()
	sumaTotal += resultado
	return sumaTotal * h
}

func main() {
	a := 2.0
	b := 20.0
	n := int64(1_000_000)
	numHilos := 4

	funcion := FuncionCuadratica{}
	integrador := IntegradorTrapecioParalelo{
		a:        a,
		b:        b,
		n:        n,
		numHilos: numHilos,
		funcion:  funcion,
	}

	inicio := time.Now()
	resultado := integrador.Integrar()
	duracion := time.Since(inicio)

	fmt.Printf("Integral aproximada (Trapecio paralelo) = %.15f\n", resultado)
	fmt.Printf("Subintervalos = %d, Hilos = %d, Tiempo = %.3f ms\n", n, numHilos, float64(duracion.Microseconds())/1000.0)
}

File: Lab04/trapecioPool.cpp
#include <iostream>
#include <cmath>
#include <thread>
#include <chrono>
using namespace std;

double f(double x){
    return 2.0 * x * x + 3.0 * x + 0.5;
}

void integrarParcial(double a, double h, long inicio, long fin, long double &resultado_parcial) {
    long double suma = 0.0;
    for(long i = inicio; i <= fin; i++){
        double x = a + i * h;
        suma += f(x);
    }
    resultado_parcial = suma;
}

long double integrarPorTrapecioParalelo(double a, double b, long n, int num_hilos) {
    double h = (b - a) / n;
    long double suma_total = 0.5L * (f(a) + f(b));

    thread* hilos = new thread[num_hilos];
    long double* resultados = new long double[num_hilos];

    long pasos_por_hilo = n / num_hilos;

    for (int i = 0; i < num_hilos; i++) {
        long inicio = i * pasos_por_hilo + 1;
        long fin = (i == num_hilos - 1) ? (n - 1) : ((i + 1) * pasos_por_hilo);
        hilos[i] = thread(integrarParcial, a, h, inicio, fin, ref(resultados[i]));
    }

    for (int i = 0; i < num_hilos; i++) {
        hilos[i].join();
        suma_total += resultados[i];
    }

    delete[] hilos;
    delete[] resultados;

    return suma_total * h;
}

int main() {
    double a = 2.0, b = 20.0;
    long n = 1000000;
    int num_hilos = 4;

    auto inicio = chrono::high_resolution_clock::now();
    long double resultado = integrarPorTrapecioParalelo(a, b, n, num_hilos);
    auto fin = chrono::high_resolution_clock::now();

    chrono::duration<double, milli> duracion = fin - inicio;

    cout.precision(15);
    cout << "Integral aproximada (Trapecio paralelo) = " << (double)resultado << endl;
    cout << "Subintervalos = " << n << ", Hilos = " << num_hilos
         << ", Tiempo = " << duracion.count() << " ms" << endl;

    return 0;
}
